### ISSUE TITLE
.NET MAUI iOS target publishing/archiving

### ISSUE THREAD

**Issue by jfversluis** [dotnet/maui#4397]
### TL;DR

* Archiving and publishing a .NET MAUI iOS app (both .NET MAUI and .NET MAUI Blazor) to an ipa works, I have been able to sign the binary. Additionally I could upload the ipa to the App Store, put it through Testflight, download and run it successfully. See the steps below.
* This is all through the command-line, UI support in Visual Studio is not there (yet) 

### Description

I have been trying to see if publishing an iOS app through .NET MAUI works at the moment. This description assumes that you have basic knowledge of publishing to the Apple App Store. 

## 1. Prerequisites

* All the required tools and bits are installed to create and run .NET MAUI iOS apps successfully, both on your Windows machine as well as a [macOS build host](https://docs.microsoft.com/xamarin/ios/get-started/installation/windows/connecting-to-mac/)
  * That includes connecting to a macOS build host
  * You will have need to connected to the macOS build host before, in which case your password is cached, or you need to know the credentials to the user account that is used on your macOS build host that is used to build your app
  * In addition to the credentials you will need the IP address of your macOS build host
* Make sure that your `info.plist` file contains a valid version number and package identifier (i.e. `com.mycompany.app`)

## 2. Gather Signing Requirements

If you currently have a certificate and provisioning profile that you can use, for example when you are updating an existing app, you can skip this step. However, you do need to know the name of the certificate in the Keychain app. See Find Certificate Name below. And you'll need to know the name of the provisioning profile, see Creating a Distribution Profile below.

These can also be obtained from your current project, probably in the iOS project csproj file.

## 2.1 Creating a Distribution Certificate

This process is no different between Xamarin, Xamarin.Forms, .NET MAUI or even a native app. Follow the steps in the [current doc](https://docs.microsoft.com/xamarin/ios/deploy-test/app-distribution/app-store-distribution/#creating-a-distribution-certificate) we have about this, under "Creating a Distribution Certificate".

### 2.2 Find Certificate Name

After following the guide above, and the certificate is installed, make sure to open the Keychain app and localize the certificate you just added. In the screenshot below the name would be `iPhone Distribution: Gerald Versluis (AY5KBJ6RN9)`. You will need this to sign your ipa file.

![image](https://user-images.githubusercontent.com/939291/151661044-f94456fb-b613-446b-8bcd-1a5509861f69.png)

### 2.3 Creating a Distribution Profile

Next thing we need is a distribution profile. Like the distribution certificate, this process is no different for Xamarin, Xamarin.Forms, .NET MAUI or a native app.  Follow the steps in the [current doc](https://docs.microsoft.com/xamarin/ios/deploy-test/app-distribution/app-store-distribution/?tabs=macos#creating-a-distribution-profile) we have about this, under "Creating a Distribution Profile".

Make sure that with step 6, where you name your distribution profile, note the name down. You will need that to sign your ipa file. In the screenshot below the name would be `iosmauisigning`.

![image](https://user-images.githubusercontent.com/939291/151661351-1e42706b-71c3-4354-a2c2-4702a4766840.png)

### 3.1 Add a (dummy) `Entitlements.plist` File (Optional)

At the time of writing the default templates do not have a `Entitlements.plist` file, this could be needed depending on what iOS features you're using, see [this doc](https://docs.microsoft.com/xamarin/ios/deploy-test/provisioning/entitlements) for more info. Add a new file under `Platforms/iOS/Entitlements.plist`.

If you don't have a specific need for it, fill it with the default contents down below. If you have an existing app, you can copy it from there as well.

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
</dict>
</plist>
```

### 3.2 Update Your csproj File With Signing Information

We need to add the information we got from above into our csproj file. If you have an existing app, you can copy this from your iOS csproj file. The information will look like this, and needs to be contained within the `<Project></Project>` node

```xml
<PropertyGroup Condition="$(TargetFramework.Contains('-ios')) and '$(Configuration)' == 'Release'">
	<RuntimeIdentifier>ios-arm64</RuntimeIdentifier>
        <!-- Only needed when you have a need for entitlements -->
	<!-- <CodesignEntitlement>Entitlements.plist</CodesignEntitlement> -->
	<CodesignKey>iPhone Distribution: Gerald Versluis (AY5KBJ6RN9)</CodesignKey>
	<CodesignProvision>iosmauisigning</CodesignProvision>
</PropertyGroup>
```

This information will only be added if you are building the iOS target and building the release configuration. Tweak this as needed.

I think all the info speaks for itself, the runtime identifier is a fixed value, all the others are values we got from the steps above.

### 3.3 Build and Sign Your App

Navigate to the folder that holds the source for your .NET MAUI iOS app,  and execute the following command:

`dotnet publish -f:net6.0-ios -c:Release /p:ServerAddress={macOS build host IP address} /p:ServerUser={macOS username} /p:ServerPassword={macOS password} /p:TcpPort=58181 /p:ArchiveOnBuild=true /p:_DotNetRootRemoteDirectory=/Users/{macOS username}/Library/Caches/Xamarin/XMA/SDKs/dotnet/`

A few things to note here:
* Replace the IP address in `/p:ServerAddress`
* Replace the username in `/p:ServerUser`
* Either replace the password in `/p:ServerPassword` or leave out that switch altogether. If you have connected with your macOS build host before, your password is cached and it shouldn't be needed.
* The `/p:_DotNetRootRemoteDirectory` option seems to be required. Note that you need to replace the username in the path.

An example of this command could look like:

` dotnet publish -f:net6.0-ios -c:Release /p:ServerAddress=192.168.1.77 /p:ServerUser=jfversluis /p:TcpPort=58181 /p:ArchiveOnBuild=true /p:_DotNetRootRemoteDirectory=/Users/jfversluis/Library/Caches/Xamarin/XMA/SDKs/dotnet/ /p:EnableAssemblyILStripping=false /p:ServerPassword=D0tN€T-M@uI-iS-@w€s0mE`

After this command had completed successfully you should see your App Store ready ipa file under `bin\Release\net6.0-ios\ios-arm64\publish`.

Note: during building I got a popup on my macOS machine asking for permission to run `codesign`. Make sure to have a look at your macOS machine if the process seems to take really long/fail. I suspect clicking "Always Allow" should prevent that from happening with future builds.

![image](https://user-images.githubusercontent.com/939291/151661999-c7e56827-0176-4c2f-bcb1-70654c07b1d5.png)

I was able to upload this through the Transporter app, make it available through Testflight and download and run the app successfully

![image](https://user-images.githubusercontent.com/939291/151662079-5c748cf8-279b-4df9-8570-6d9da3bef57a.png)

## Notes
* Instead of specifying all the options with the `dotnet publish` command you can also add all of those to your csproj. The part after `/p:` would be the node name. E.g. `/p:ArchiveOnBuild=true` would be `<ArchiveOnBuild>true</ArchiveOnBuild>`
* In extend to that, the options that I added to the csproj file can also be options in the `dotnet publish` command. E.g. `<CodesignProvision>iosmauisigning</CodesignProvision>` would become `/p:CodesignProvision=iosmauisigning`.

## Related Resources
* https://docs.microsoft.com/xamarin/ios/deploy-test/app-distribution/app-store-distribution/
* https://docs.microsoft.com/xamarin/ios/deploy-test/provisioning/entitlements
* https://stackoverflow.com/questions/70437037/net-6-publishing-android-application

## Also See
* #4329
* #4377

**Comment by ElliottBrand** [comment]
First, this is amazing and the kind of documentation I've been looking for in regards to MAUI iOS archiving! Some people who aren't so used to native development (like me) may not know what the IP Address to the Mac is. I happened to find it in the popup when I clicked the "Pair to Mac" icon.

Once I initialize everything, the provisioning works fine and things start running, but it hangs up and crashes for me after the linker is running.

``` 
Optimizing assemblies for size, which may change the behavior of the app. Be sure to test after publishing. See: https://aka.ms/dotnet-illink
  Tool xcrun execution finished (exit code = 1).

  Undefined symbols for architecture arm64:
    "_OBJC_CLASS_$_FIRAnalytics", referenced from:
        objc-class-ref in registrar.o
       (maybe you meant: _OBJC_CLASS_$_FIRAnalyticsConfiguration)
  ld: symbol(s) not found for architecture arm64
  clang: error: linker command failed with exit code 1 (use -v to see invocation)

C:\Program Files\dotnet\packs\Microsoft.iOS.Sdk\15.2.200-preview.12.4\targets\Xamarin.Shared.Sdk.targets(1089,3): error : clang++ exited with code 1
```

**Comment by xperiandri** [comment]
Adding `net6.0-macos` to `TargetFrameworks` breaks the build logic and redirects iOS build to macOS SDK

**Comment by xperiandri** [comment]
Only doing this
``` XML
    <TargetFrameworks>net6.0-android;net6.0-ios;net6.0-macos</TargetFrameworks>
    <TargetFrameworks Condition="'$(_PlatformName)' == 'Android'">net6.0-android</TargetFrameworks>
    <TargetFrameworks Condition="'$(_PlatformName)' == 'iOS'">net6.0-ios</TargetFrameworks>
    <TargetFrameworks Condition="'$(_PlatformName)' == 'macOS'">net6.0-macos</TargetFrameworks>
```
And then specifying `/p:_PlatformName=iOS` fixes the issue

**Comment by xperiandri** [comment]
Not a Mac Catalyst. I build Uno Platform app which supports macOS natively so that I don't need Mac Catalyst.
I commented out that target.

**Comment by xperiandri** [comment]
So with RC1 I can build iOS but not publish

```
/usr/local/share/dotnet/packs/Microsoft.iOS.Sdk/15.2.303-rc.1.55/targets/Xamarin.Shared.Sdk.targets(1411,3): warning : The file '/Users/xperiandri/.nuget/packages/fsharp.core/6.0.3/contentFiles/any/netstandard2.1/FSharp.Core.xml' does not specify a 'PublishFolderType' metadata, and a default value could not be calculated. The file will not be copied to the app bundle. [/Users/xperiandri/Projects/ITM/Client/Client.Mobile/Transmission.Client.Mobile.csproj]
/usr/local/share/dotnet/packs/Microsoft.iOS.Sdk/15.2.303-rc.1.55/targets/Xamarin.Shared.Sdk.targets(1411,3): warning : The file '/Users/xperiandri/Projects/ITM/Client/Client/bin/Release/netstandard2.0/Transmission.Client.Programs.xml' does not specify a 'PublishFolderType' metadata, and a default value could not be calculated. The file will not be copied to the app bundle. [/Users/xperiandri/Projects/ITM/Client/Client.Mobile/Transmission.Client.Mobile.csproj]
/usr/local/share/dotnet/packs/Microsoft.iOS.Sdk/15.2.303-rc.1.55/targets/Xamarin.Shared.Sdk.targets(1411,3): warning : The file '/Users/xperiandri/Projects/ITM/Shared/Domain/bin/Release/netstandard2.0/Transmission.Domain.xml' does not specify a 'PublishFolderType' metadata, and a default value could not be calculated. The file will not be copied to the app bundle. [/Users/xperiandri/Projects/ITM/Client/Client.Mobile/Transmission.Client.Mobile.csproj]
/usr/local/share/dotnet/packs/Microsoft.iOS.Sdk/15.2.303-rc.1.55/targets/Xamarin.Shared.Sdk.targets(1411,3): warning : The file '/Users/xperiandri/Projects/ITM/Shared/FSharp/bin/Release/netstandard2.0/Transmission.FSharp.xml' does not specify a 'PublishFolderType' metadata, and a default value could not be calculated. The file will not be copied to the app bundle. [/Users/xperiandri/Projects/ITM/Client/Client.Mobile/Transmission.Client.Mobile.csproj]
ILLink : error IL2035: System.Net.Http.HttpClientHandler.GetClientCertificateOptions(): Unresolved assembly 'Xamarin.iOS' in 'DynamicDependencyAttribute'. [/Users/xperiandri/Projects/ITM/Client/Client.Mobile/Transmission.Client.Mobile.csproj]
ILLink : error IL2035: System.Net.Http.HttpClientHandler.GetClientCertificates(): Unresolved assembly 'Xamarin.iOS' in 'DynamicDependencyAttribute'. [/Users/xperiandri/Projects/ITM/Client/Client.Mobile/Transmission.Client.Mobile.csproj]
ILLink : error IL2035: System.Net.Http.HttpClientHandler.SetClientCertificateOptions(ClientCertificateOption): Unresolved assembly 'Xamarin.iOS' in 'DynamicDependencyAttribute'. [/Users/xperiandri/Projects/ITM/Client/Client.Mobile/Transmission.Client.Mobile.csproj]
ILLink : error IL2035: System.Net.Http.HttpClientHandler.SetCredentials(ICredentials): Unresolved assembly 'Xamarin.iOS' in 'DynamicDependencyAttribute'. [/Users/xperiandri/Projects/ITM/Client/Client.Mobile/Transmission.Client.Mobile.csproj]
ILLink : error IL2035: System.Net.Http.HttpClientHandler.SetMaxConnectionsPerServer(Int32): Unresolved assembly 'Xamarin.iOS' in 'DynamicDependencyAttribute'. [/Users/xperiandri/Projects/ITM/Client/Client.Mobile/Transmission.Client.Mobile.csproj]
ILLink : error IL2035: System.Net.Http.HttpClientHandler.SetProxy(IWebProxy): Unresolved assembly 'Xamarin.iOS' in 'DynamicDependencyAttribute'. [/Users/xperiandri/Projects/ITM/Client/Client.Mobile/Transmission.Client.Mobile.csproj]
  Optimizing assemblies for size, which may change the behavior of the app. Be sure to test after publishing. See: https://aka.ms/dotnet-illink
/usr/local/share/dotnet/sdk/6.0.300-preview.22154.4/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ILLink.targets(109,5): error NETSDK1144: Optimizing assemblies for size failed. Optimization can be disabled by setting the PublishTrimmed property to false. [/Users/xperiandri/Projects/ITM/Client/Client.Mobile/Transmission.Client.Mobile.csproj]
```

**Comment by softlion** [comment]
This is xamforms; not maui. But same codebase i suppose. VS 2022 preview 2.1
```
PathTooLongException: The path '/Users/mac3/Library/Caches/Xamarin/mtbs/builds/Wz.iOS/dbe16c3bde3d529ebc2e45d0e69b86efed9380ddc5acdf2a2d42ba8b696fed84/obj\iPhone\Debug\device-builds\iphone11.8-15.3.1/codesign\\bin\iPhone\Debug\device-builds\iphone11.8-15.3.1\WziOS.app\\Frameworks\libSkiaSharp.framework\libSkiaSharp.Substring(bin\iPhone\Debug\device-builds\iphone11.8-15.3.1\WziOS.app\\Frameworks\libSkiaSharp.framework\libSkiaSharp.LastIndexOf('.app' is too long, or a component of the specified path is too long.
```

**Comment by xperiandri** [comment]
```
xperiandri@Andrii-Mac Client.Mobile % dotnet publish -f net6.0-macos -r osx-x64 -c Release /p:_PlatformName=macOS --no-restore
Microsoft (R) Build Engine version 17.2.0-preview-22126-01+40f49cd8d for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
/usr/local/share/dotnet/sdk/6.0.300-preview.22154.4/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(1110,5): warning NETSDK1179: One of '--self-contained' or '--no-self-contained' options are required when '--runtime' is used. [/Users/xperiandri/Projects/ITM/Client/Client.Mobile/Transmission.Client.Mobile.csproj]
  Detected signing identity:
          
    Bundle Id: com.transmission.baggage
    App Id: com.transmission.baggage
/Users/xperiandri/Projects/ITM/Client/Client.Mobile/obj/Release/net6.0-macos/osx-x64/actool/cloned-assets/Assets.xcassets : actool warning : The image set "unologo" has 8 unassigned children. [/Users/xperiandri/Projects/ITM/Client/Client.Mobile/Transmission.Client.Mobile.csproj]
/Users/xperiandri/Projects/ITM/Client/Client.Mobile/obj/Release/net6.0-macos/osx-x64/actool/cloned-assets/Assets.xcassets : actool warning : The image set "unologo" has 8 unassigned children. [/Users/xperiandri/Projects/ITM/Client/Client.Mobile/Transmission.Client.Mobile.csproj]
/usr/local/share/dotnet/sdk/6.0.300-preview.22154.4/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(267,5): error NETSDK1047: Assets file '/Users/xperiandri/Projects/ITM/Client/Client.Mobile/obj/project.assets.json' doesn't have a target for 'net6.0-macos/osx-x64'. Ensure that restore has run and that you have included 'net6.0-macos' in the TargetFrameworks for your project. You may also need to include 'osx-x64' in your project's RuntimeIdentifiers. [/Users/xperiandri/Projects/ITM/Client/Client.Mobile/Transmission.Client.Mobile.csproj]
```

**Comment by xperiandri** [comment]
```
xperiandri@Andrii-Mac Client.Mobile % dotnet publish -f net6.0-macos -c Release /p:_PlatformName=macOS --no-restore 
Microsoft (R) Build Engine version 17.2.0-preview-22126-01+40f49cd8d for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  Detected signing identity:
          
    Bundle Id: 
    App Id: 
/usr/local/share/dotnet/sdk/6.0.300-preview.22154.4/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(267,5): error NETSDK1047: Assets file '/Users/xperiandri/Projects/ITM/Client/Client.Mobile/obj/project.assets.json' doesn't have a target for 'net6.0-macos/osx-x64'. Ensure that restore has run and that you have included 'net6.0-macos' in the TargetFrameworks for your project. You may also need to include 'osx-x64' in your project's RuntimeIdentifiers. [/Users/xperiandri/Projects/ITM/Client/Client.Mobile/Transmission.Client.Mobile.csproj]
```

**Comment by xperiandri** [comment]
```xperiandri@Andrii-Mac Client.Mobile % dotnet build -f net6.0-ios -r ios-arm64 -c Release /p:_PlatformName=iOS /p:ArchiveOnBuild=true --no-restore --self-contained
Microsoft (R) Build Engine version 17.2.0-preview-22126-01+40f49cd8d for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  Detected signing identity:
          
    Code Signing Key: "Apple Development: Konstantin Persisty (J776Q7ZSS6)" (5549BFE4DC9A10A6AF4634B140C29640DFB42B6C)
    Provisioning Profile: "VS: WildCard Development" (40665c5a-c3a4-4069-ad2d-2cb03a1e3695)
    Bundle Id: com.transmission.baggage
    App Id: 2NZ5U9QKL5.com.transmission.baggage
/usr/local/share/dotnet/sdk/6.0.300-preview.22154.4/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(267,5): error NETSDK1047: Assets file '/Users/xperiandri/Projects/ITM/Client/Client.Mobile/obj/project.assets.json' doesn't have a target for 'net6.0-ios/ios-arm64'. Ensure that restore has run and that you have included 'net6.0-ios' in the TargetFrameworks for your project. You may also need to include 'ios-arm64' in your project's RuntimeIdentifiers. [/Users/xperiandri/Projects/ITM/Client/Client.Mobile/Transmission.Client.Mobile.csproj]

Build FAILED.

/usr/local/share/dotnet/sdk/6.0.300-preview.22154.4/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(267,5): error NETSDK1047: Assets file '/Users/xperiandri/Projects/ITM/Client/Client.Mobile/obj/project.assets.json' doesn't have a target for 'net6.0-ios/ios-arm64'. Ensure that restore has run and that you have included 'net6.0-ios' in the TargetFrameworks for your project. You may also need to include 'ios-arm64' in your project's RuntimeIdentifiers. [/Users/xperiandri/Projects/ITM/Client/Client.Mobile/Transmission.Client.Mobile.csproj]
    0 Warning(s)
    1 Error(s)
    ```

**Comment by mcostainfo** [comment]
I had this problem, can anyone help me,

  /usr/bin/mdimport bin/Release/net6.0-ios/ios-arm64/
  Tool /usr/bin/codesign execution finished (exit code = 1).

  Failed to parse entitlements: AMFIUnserializeXML: syntax error near line 22
![Uploading Captura de tela 2022-04-30 115019.png…]()

**Comment by xperiandri** [comment]
@mcostainf which version?

**Comment by mcostainfo** [comment]
Hi xperiandri , Version 17.2.0 Preview 5.0

**Comment by xperiandri** [comment]
Hi. Have you done `dotnet workload update`?
Are you on MAUI or Uno?

**Comment by mcostainfo** [comment]
@xperiandri  MAUI

**Comment by xperiandri** [comment]
Try `dotnet workload update` on both Windows and macOS
If does not help, try `maui-check` .NET tool

**Comment by mcostainfo** [comment]
@xperiandri  Another thing I'm using MAC on "macincloud.com" , 
I managed to get past that error, now this error:
![Captura de tela 2022-04-30 124242](https://user-images.githubusercontent.com/53783760/166112454-60c1df3b-5b32-462d-ab6f-ba0fb47f9817.png)

**Comment by xperiandri** [comment]
Maybe you can specify the resulting folder in Visual Studio UI?

**Comment by mcostainfo** [comment]
@xperiandri Thanks for your collaboration.
@jfversluis I would like to thank jfversluis , the contents of are excellent.

**Comment by otium99** [comment]
@jfversluis  I have followed the official docs and watched your video on it, but no matter how I turn it, it just won't publish:

https://stackoverflow.com/questions/72399444/errors-with-dotnet-publish

PS: thanks for the video series though, that's really helpful to get started.

**Comment by JeffIPV** [comment]
@jfversluis  I also have the same issues as @otium99 . I got it to work once in rc2 after trying 10 times. Doing the exact same thing in his stackoverflow
But can no longer get it to work in rc2 or rc3 workload.
